tests/suite/graceful_recovery_test.go

```diff
@@ -0,0 +1,316 @@
+
+	// propagation policy is set to delete underlying pod created through job
+	Expect(resourceManager.Delete(
+		[]client.Object{job},
+		client.PropagationPolicy(metav1.DeletePropagationBackground),
+	)).ToNot(HaveOccurred())
+}
+
+func waitForContainerRestart(ngfPodName, containerName string, currentRestartCount int) error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeoutConfig.RequestTimeout)
+	defer cancel()
+
+	//nolint:nilerr
+	return wait.PollUntilContextCancel(
+		ctx,
+		500*time.Millisecond,
+		true, /* poll immediately */
+		func(_ context.Context) (bool, error) {
+			restartCount, err := getContainerRestartCount(containerName, ngfPodName)
+			if err != nil {
+				return false, nil
+			}
+
+			return restartCount == currentRestartCount+1, nil
+		},
+	)
+}
+
+func waitForWorkingTraffic() error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeoutConfig.RequestTimeout)
+	defer cancel()
+
+	//nolint:nilerr
+	return wait.PollUntilContextCancel(
+		ctx,
+		500*time.Millisecond,
+		true, /* poll immediately */
+		func(_ context.Context) (bool, error) {
+			if err := expectRequestToSucceed(teaURL, address, "URI: /tea"); err != nil {
+				return false, nil
+			}
+			if err := expectRequestToSucceed(coffeeURL, address, "URI: /coffee"); err != nil {
+				return false, nil
+			}
+			return true, nil
+		},
+	)
+}
+
+func waitForFailingTraffic() error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeoutConfig.RequestTimeout)
+	defer cancel()
+
+	//nolint:nilerr
+	return wait.PollUntilContextCancel(
+		ctx,
+		500*time.Millisecond,
+		true, /* poll immediately */
+		func(_ context.Context) (bool, error) {
+			if err := expectRequestToFail(teaURL, address, "URI: /tea"); err != nil {
+				return false, nil
+			}
+			if err := expectRequestToFail(coffeeURL, address, "URI: /coffee"); err != nil {
+				return false, nil
+			}
+			return true, nil
+		},
+	)
+}
+
+func expectRequestToSucceed(appURL, address string, responseBodyMessage string) error {
+	status, body, err := framework.Get(appURL, address, timeoutConfig.RequestTimeout)
+	if status != http.StatusOK {
+		return errors.New("http status was not 200")
+	}
+
+	if !strings.Contains(body, responseBodyMessage) {
+		return errors.New("expected response body to contain correct body message")
+	}
+
+	return err
+}
+
+func expectRequestToFail(appURL, address string, responseBodyMessage string) error {
+	status, body, err := framework.Get(appURL, address, timeoutConfig.RequestTimeout)
+	if status != 0 {
+		return errors.New("expected http status to be 0")
+	}
+
+	if strings.Contains(body, responseBodyMessage) {
+		return errors.New("expected response body to not contain correct body message")
+	}
+
+	if err == nil {
+		return errors.New("expected request to error")
+	}
+
+	return nil
+}
+
+func checkContainerLogsForErrors(ngfPodName string) {
+	sinceSeconds := int64(15)
```

Comment:
how did you come up with 15 seconds value?

Reply:
This was completely arbitrary. Do you recommend something different? Also do you recommend moving this variable out into some constant?